### PR TITLE
feat(hooks-store): new design & refactoring

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Design: https://www.figma.com/design/1ctikx9sFa9Ff925GRsjvC/Hook-Store

Changes:
1. Separated pure/container components
2. Simplified `HookDappContext`: now it is passed as a component prop instead of react context
3. Added localStorage persisting for hooks state
4. Added hooks editing
5. And of course, applied the new design from [Michel's PR](https://github.com/cowprotocol/cowswap/pull/4852) 

| -- | -- | -- | -- | -- |
|--------|--------|--------|--------|--------|
| image | image | image | image | image | 

# To Test

Please, keep in mind that the feature is not ready yet and still under feature-flag.

1. UI should corrsepond to the Figma layouts. (reordering feature is not implemented yet)
2. Added hooks should be saved in localStorage (should be applied to the form after page refresh)
3. "Build your own" hook should be editable

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/hooksStore/state/hookDetailsAtom.ts [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
+  preHooks: CowHookDetailsSerialized[]
+  postHooks: CowHookDetailsSerialized[]
+}
 
-export const hooksAtom = atom<{ preHooks: CowHookDetails[]; postHooks: CowHookDetails[] }>({
-  preHooks: [],
-  postHooks: [],
-})
+export const hooksAtom = atomWithStorage<HooksStoreState>(
+  'hooks-store-atom:v1',
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/hooksStore/state/hookDetailsAtom.ts:12]: camelCase

```suggestion
  'hooksStoreAtom:v0',
```

**Review by alfetopito — COMMENTED** [review]
Minor comments did not test it.

Need to make sure permit is tested with and without hooks to not break it again.

**Comment by elena-zh** [comment]
Hey @shoom3301 , great job!
However, I have some issues and questions: 
1. Hooks will not be used for wraps/unwraps, so I'd delete them from the form in these cases: 
image
2. If a hook is specified in one network, it should be removed from the form once a user changes a network (or connects to a wallet in another network)
3. I've added hooks on the Hooks tab, and then decided to do a regular swap: hooks were added to the appdata (IMO, hooks should be added to the appdata only when an order is placed from the Hooks tab)
4. I think, it would be great not to allow to add the same hook twice into the pre- or post-hook section
image
5. Same case as for the one above: it would be great to add a validation when a user fills in the same custom hook (with the same data in all fields)
6. If 2 different custom hooks were added, it would be great to distinguish them somehow. Maybe add a hooks fields preview on a hook hover? 
image
7. A specified hook was not added to the appdata for a ETH-flow order
8. A specified hook was not added to the appdata in Safe
9. If I add several hooks, and run simulation for each of them, there will be only one (the latest) link with a simulation report. Is it possible to keep them all on the form? 
щтдн щту
10. (weird) in this PR I was not able to connect to Impersonator wallet using WC
wwc
11. Hook validation: IMO, it would be great to disable a possibility of adding a hook when there is nothing to claim: 
image
12. UI things (for @fairlighteth ): 
- there should not be 'details' and version on a hook 
UI
- there should be search for a  hook option
- it would be great to display nicer 'connect to a wallet' message on a hook details
connect wallet first
- it would be great to improve the UI on a reduced screen 
image


Thanks!

**Comment by shoom3301** [comment]
@elena-zh thank you!

1. Fixed
2. Fixed, now the hooks state is bound to chainId + account
3. Fixed
4. Out of this task scope
5. Out of this task scope
6. As a fast solution I added title which displays a hook uid
image
7. It is intended
8. It is intended
9. Out of this task scope
10. Couldn't reproduce
11. Out of this task scope
12. I don't think so
13. Out of this task scope

**Comment by shoom3301** [comment]
@elena-zh "Out of this task scope" - means I don't think we should spend time on this right now, I would reiterate this after we get some feedback from users

**Comment by elena-zh** [comment]
Hey @shoom3301 , thank you for explanation!

Some comments: 
1. If a hook should not be applied to ETH order, IMO, it would be hide it from the form once ETH (XDAI) token is selected as a sell one. 
2. Safe or other SC wallets: this should be more global change: if a hook store should not work with SC wallets, it would be nice to hide this tab when connected to it (not sure!). However, need to discuss it with the team as [here](https://cowservices.slack.com/archives/C0361CDG8GP/p1721235010655299?thread_ts=1721212169.951999&cid=C0361CDG8GP) it was said that hooks may work with SC wallets... 
3. Hook may not be  saved for an account: 
 - connect to a wallet
 - specify a hook
 - change a wallet
 - add another hook for another account
 - change a wallet to the 1st address --> the hook for that address is not saved. 

Thanks!

**Comment by shoom3301** [comment]
@elena-zh it's a bit complicated for now, let's keep it as it is (hooks are always disabled for eth-flow)

**Comment by elena-zh** [comment]
> @elena-zh it's a bit complicated for now, let's keep it as it is (hooks are always disabled for eth-flow)

added this case into #4662

**Review by elena-zh — COMMENTED** [review]
Hey @shoom3301, there are several UI and low priority issues, but I will open a separate task to address them. 
But I'd like to mention one issue that I'm concerned about: 
1. Place an order with a permittable token (not approved) through Hooks form
2. While this order is Open, go to the swap page and place an order with the same token pair.

**AR**: there are 2 possible outcomes: 
1. The app requires me to sign a permit request once again
2. The app does not allow me to place the order due to 'insufficient allowance' error
4. in both cases   I see error fetching permit  data on console. 

permit
куйгшкуы

Could you please double check this case and LMK whether this should be addressed in the current PR or in the another task? 
Thanks

**Comment by shoom3301** [comment]
@elena-zh thanks! Please put it in a new task, we will take care of it later

**Comment by elena-zh** [comment]
> @elena-zh thanks! Please put it in a new task, we will take care of it later

Now reported here: #4727 

All the rest issues can be found in #4662

**Review by elena-zh — APPROVED** [review]

**Pull request merged**